### PR TITLE
Hocs

### PR DESCRIPTION
**Pull request by fsih**
### Resolves
In response to https://github.com/LLK/scratch-paint/pull/623#discussion_r213440398

### Proposed Changes
Break out a lot of HOCs from paint-editor (move code out of the salad that is paint-editor)

### Reason for Changes
Better encapsulation in code
No changes in functionality

### Testing
See https://llk.github.io/scratch-gui/hocs/

### CODE REVIEW

File: src/hocs/undo-hoc.jsx [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+                'handleUndo',
+                'handleRedo',
+                'handleSetSelectedItems',
+                'shouldShowUndo',
+                'shouldShowRedo'
+            ]);
+        }
+        handleUndo () {
+            performUndo(this.props.undoState, this.props.onUndo, this.handleSetSelectedItems, this.props.onUpdateImage);
+            // TODO this.handleUpdateImage
```
Comment (thisandagain) [inline src/hocs/undo-hoc.jsx]: Do you have a Github issue filed for this? If not, would you mind filing an issue and linking it to this `@todo` item?

File: src/hocs/update-image-hoc.jsx [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+                performSnapshot(this.props.undoSnapshot, Formats.BITMAP);
+            }
+        }
+        handleUpdateVector (skipSnapshot) {
+            const guideLayers = hideGuideLayers(true /* includeRaster */);
+
+            // Export at 0.5x
+            scaleWithStrokes(paper.project.activeLayer, .5, new paper.Point());
+            const bounds = paper.project.activeLayer.bounds;
+            // @todo generate view box
```
Comment (thisandagain) [inline src/hocs/update-image-hoc.jsx]: Same here.

**Review by thisandagain — DISMISSED** [review]
This looks really clean @fsih! I did some testing in the playground as well as the branch you pushed and everything is looking good.

File: src/hocs/undo-hoc.jsx (hunk 1 above)
Comment (fsih) [reply src/hocs/undo-hoc.jsx]: Oops, that was just a note to self that's already been fixed

File: src/hocs/update-image-hoc.jsx (hunk 2 above)
Comment (fsih) [reply src/hocs/update-image-hoc.jsx]: Fixed

**Pull request merged**